pintc/src/types.rs

```diff
@@ -483,7 +483,10 @@ impl Type {
         match self {
             Type::Map { ty_from, ty_to, .. } => {
                 // We only support maps from these types for now
-                (ty_from.is_bool() || ty_from.is_int() || ty_from.is_b256())
+                (ty_from.is_bool()
+                    || ty_from.is_int()
+                    || ty_from.is_b256()
+                    || ty_from.is_allowed_in_storage())
```

Comment:
So this is a bit incorrect since it would allow `ty_from` to be a map and that's never going to be allowed. I think I misled you in the issue. Instead, 

I think we can use the method `get_storage_only_ty` which, if returns `None`, then we know that the type can be used as a key. Then, we add tests for each case:
1. all primitive types.
2. tuples
3. unions
4. arrays
All as long as they don't _contain_ storage only types (which are basically maps and vectors only)